implementations/rust/ockam/ockam/src/credential.rs

```diff
@@ -1,44 +1,270 @@
+#![allow(missing_docs)]
+
 //! Building block for Attribute-based access control
 
 mod exchange;
 
-pub use exchange::*;
-
+use core::time::Duration;
+use minicbor::bytes::ByteSlice;
 use minicbor::{Decode, Encode};
-use ockam_core::CowBytes;
+use ockam_core::compat::borrow::Cow;
+use ockam_core::compat::collections::BTreeMap;
+use ockam_core::errcode::{Kind, Origin};
+use ockam_core::vault::{Signature, SignatureVec};
+use ockam_core::{CowBytes, CowStr, Error, Result};
+use ockam_identity::{Identity, IdentityIdentifier, IdentityVault};
+
+pub use exchange::*;
 
 #[cfg(feature = "tag")]
 use crate::TypeTag;
 
-/// Credential
+pub const MAX_CREDENTIAL_VALIDITY: Duration = Duration::from_secs(6 * 3600);
+
 #[derive(Debug, Decode, Encode)]
 #[rustfmt::skip]
 #[cbor(map)]
 pub struct Credential<'a> {
     #[cfg(feature = "tag")]
     #[n(0)] tag: TypeTag<3796735>,
+    /// CBOR-encoded [`CredentialData`].
     #[b(1)] attributes: CowBytes<'a>,
+    /// Cryptographic signature of attributes data.
     #[b(2)] signature: CowBytes<'a>,
 }
 
+#[derive(Debug, Decode, Encode)]
+#[rustfmt::skip]
+#[cbor(map)]
+pub struct CredentialData<'a> {
+    /// A schema identifier to allow distinguishing sets of attributes.
+    #[n(1)] schema: Option<SchemaId>,
+    /// User-defined key-value pairs.
+    #[b(2)] attributes: Option<Attributes<'a>>,
+    /// The subject this credential is issued for.
+    #[n(3)] subject: IdentityIdentifier,
+    /// The entity that signed this credential.
+    #[n(4)] issuer: IdentityIdentifier,
+    /// The label of the issuer's public key.
+    #[b(5)] issuer_key: CowStr<'a>,
+    /// The time when this credential was created.
+    #[n(6)] created: Timestamp,
+    /// The time this credential expires.
+    #[n(7)] expires: Timestamp
+}
+
 impl<'a> Credential<'a> {
-    /// Constructor
-    pub fn new(attributes: CowBytes<'a>, signature: CowBytes<'a>) -> Self {
-        Credential {
-            #[cfg(feature = "tag")]
-            tag: TypeTag,
-            attributes,
-            signature,
+    pub fn builder<'b>(subject: IdentityIdentifier) -> CredentialBuilder<'b> {
+        CredentialBuilder {
+            schema: None,
+            subject,
+            attrs: Attributes::new(),
+            validity: MAX_CREDENTIAL_VALIDITY,
         }
     }
 
-    /// Attributes getter
     pub fn attributes(&self) -> &[u8] {
         &self.attributes
     }
 
-    /// Signature getter
     pub fn signature(&self) -> &[u8] {
         &self.signature
     }
+
+    /// Perform a signature check with the given identity.
+    pub async fn is_signature_of<V>(&self, issuer: &Identity<V>) -> Result<bool>
```

Comment:
I would say if it is doing verification, it should have verify in its name. My first assumption from this name was that it checks if issue in the credential is the same as issuer in the argument